### PR TITLE
[Search] Add config for Azure Directory's cache path

### PR DESCRIPTION
**Pull request by loic-sharma**
By default, Lucene will store the Azure Directory cache inside of the temp path. On Azure, this is the same drive as the operating system, which only has 20GB of free space. This configuration will allow us to move the Azure Directory cache to a drive with more free space. If `AzureDirectoryCachePath` configuration isn't set, the search service will default to using the temp path.

### CODE REVIEW

File: src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs [hunk 1]
```diff
@@ -75,7 +79,24 @@ public bool Reload(IndexingConfiguration config)
...
+            {
+                sourceDirectory = new AzureDirectory(storageAccount, _indexContainerName);
+            }
+            else
+            {
+                _logger.LogInformation("Using custom azure directory cache path: {AzureDirectoryCachePath}", config.AzureDirectoryCachePath);
+
+                var azureDirectoryCachePath = Path.Combine(config.AzureDirectoryCachePath, _indexContainerName);
+
+                IODirectory.CreateDirectory(azureDirectoryCachePath);
```
Comment (agr) [inline src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs]: What will happen if path exists?

File: src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs [hunk 2]
```diff
@@ -75,7 +79,24 @@ public bool Reload(IndexingConfiguration config)
...
-            var sourceDirectory = new AzureDirectory(storageAccount, _indexContainerName);
+            AzureDirectory sourceDirectory;
+
+            if (string.IsNullOrEmpty(config.AzureDirectoryCachePath))
+            {
+                sourceDirectory = new AzureDirectory(storageAccount, _indexContainerName);
+            }
+            else
+            {
+                _logger.LogInformation("Using custom azure directory cache path: {AzureDirectoryCachePath}", config.AzureDirectoryCachePath);
```
Comment (joelverhagen) [inline src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs]: I would either also log the absolute path with `Path.GetFullPath`, pass the full path to `AzureDirectory` ctor, or require `Path.IsPathRooted`.

As it stands today, a configured relative path has unclear behavior.

File: src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs (hunk 1 above)
Comment (agr) [reply src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs]: Should we use `DirectoryInfo.Create` instead?

Comment (joelverhagen) [reply src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs]: `CreateDirectory` no-ops if the directory already exists.

Comment (loic-sharma) [reply src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs]: Nothing will happen :)

Source: https://docs.microsoft.com/en-us/dotnet/api/system.io.directory.createdirectory?view=netframework-4.7.1#System_IO_Directory_CreateDirectory_System_String_

File: src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs]: Fixed

Comment (agr) [reply src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs]: FYI, `IsPathRooted` does not guarantee path is absolute. Try
    
     [System.IO.Path]::IsPathRooted('c:a.txt');

in powershell.

Comment (joelverhagen) [reply src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs]: Oh good point. In this case, I would just avoid `IsPathRooted`...

Comment (loic-sharma) [reply src/NuGet.Indexing/IndexDirectoryProvider/CloudIndexDirectoryProvider.cs]: I went ahead and did `Path.GetFullPath`

File: src/NuGet.Services.BasicSearch/SafeRoleEnvironment.cs [hunk 3]
```diff
@@ -62,6 +62,11 @@ public static string GetDeploymentId()
             return string.Empty;
         }
 
+        public static string GetLocalResourceRootPath(string name)
+        {
+            return RoleEnvironment.GetLocalResource(name).RootPath;
```
Comment (loic-sharma) [inline src/NuGet.Services.BasicSearch/SafeRoleEnvironment.cs]: This throws if a local resource with that `name` doesn't exist.

File: src/NuGet.Services.BasicSearch/Startup.cs [hunk 4]
```diff
@@ -91,6 +91,26 @@ public class Startup
...
+            {
+                if (!SafeRoleEnvironment.IsAvailable)
+                {
+                    _logger.LogWarning(
+                        "Cannot use Azure Local Resource {LocalResourceName} for caching when the RoleEnvironment is not available",
+                        config.AzureDirectoryCacheLocalResourceName);
+                }
+                else
+                {
+                    config.AzureDirectoryCachePath = SafeRoleEnvironment.GetLocalResourceRootPath(config.AzureDirectoryCacheLocalResourceName);
```
Comment (joelverhagen) [inline src/NuGet.Services.BasicSearch/Startup.cs]: Let's document on `AzureDirectoryCachePath` that it gets overwritten at runtime if `AzureDirectoryCacheLocalResourceName` is set.

File: src/NuGet.Services.BasicSearch/SafeRoleEnvironment.cs [hunk 5]
```diff
@@ -62,6 +62,11 @@ public static string GetDeploymentId()
             return string.Empty;
         }
 
+        public static string GetLocalResourceRootPath(string name)
```
Comment (joelverhagen) [inline src/NuGet.Services.BasicSearch/SafeRoleEnvironment.cs]: We need to test what this does when:

1. Size is "small", e.g. 50 GB.
1. Size is large enough to be less than max disk size but too big given already consumed disk space, e.g. 199 GB.
1. Size is exactly what our VM SKU advertises, e.g. 200 GB.
1. Size is great than VM disk size, e.g. 300 GB.

Comment (joelverhagen) [reply src/NuGet.Services.BasicSearch/SafeRoleEnvironment.cs]: Also, what does `cleanOnRoleRecycle` do for us? What will we set it to on our resource?

**Comment by loic-sharma** [comment]
See Local Storage: https://github.com/Huachao/azure-content/blob/master/articles/cloud-services/cloud-services-configure-local-storage-resources.md

File: src/NuGet.Services.BasicSearch/Startup.cs (hunk 4 above)
Comment (loic-sharma) [reply src/NuGet.Services.BasicSearch/Startup.cs]: Fixed

**Pull request merged**